From 640488c713f60131fb1a1f00a13f630893dfb57f Mon Sep 17 00:00:00 2001
From: Suman Anna <s-anna@ti.com>
Date: Thu, 29 May 2014 17:51:48 -0500
Subject: [PATCH 15/21] remoteproc/pruss: add pru-specific debugfs support

The remoteproc core driver creates certain standard debugfs
entries, that does not give a whole lot of useful information
for the PRUs. The PRU remoteproc driver is enhanced to add
additional debugfs entries for PRU. These will be auto-cleaned
up when the parent rproc debug directory is removed.

The added debugfs support adds two new entries: 'regs' and
'single_step'. The 'regs' dumps out the useful CTRL sub-module
registers as well as 32 registers each of the GPREGs and CT_REGs.
The GPREGs and CT_REGs though are printed only when the PRU is
halted and accessible as per the IP design.

The 'single_step' utilizes the single-step execution of the PRU
cores. Writing a non-zero value performs a single step, and a
zero value restores the PRU to execute in the same as mode as
before the first single step. (note: if the PRU is halted
because of a halt instruction, then no change occurs).

Signed-off-by: Suman Anna <s-anna@ti.com>
---
 drivers/remoteproc/pruss_remoteproc.c |  132 +++++++++++++++++++++++++++++++++
 1 file changed, 132 insertions(+)

diff --git a/drivers/remoteproc/pruss_remoteproc.c b/drivers/remoteproc/pruss_remoteproc.c
index 212b976..9060174 100644
--- a/drivers/remoteproc/pruss_remoteproc.c
+++ b/drivers/remoteproc/pruss_remoteproc.c
@@ -20,6 +20,7 @@
 
 #include <linux/module.h>
 #include <linux/bitops.h>
+#include <linux/debugfs.h>
 #include <linux/of_device.h>
 #include <linux/interrupt.h>
 #include <linux/pm_runtime.h>
@@ -184,6 +185,8 @@ struct pruss {
  * @sdram_da: device address of secondary Data RAM for this PRU
  * @shrdram_da: device address of shared Data RAM
  * @fw_name: name of firmware image used during loading
+ * @dbg_single_step: debug flag to set PRU into single step mode
+ * @dbg_continous: debug flag to restore PRU execution mode
  */
 struct pru_rproc {
 	int id;
@@ -199,6 +202,8 @@ struct pru_rproc {
 	u32 sdram_da;
 	u32 shrdram_da;
 	const char *fw_name;
+	u32 dbg_single_step;
+	u32 dbg_continuous;
 };
 
 static inline u32 pruss_intc_read_reg(struct pruss *pruss, unsigned int reg)
@@ -294,6 +299,131 @@ static void *pru_i_da_to_va(struct pru_rproc *pru, u32 da, int len)
 	return NULL;
 }
 
+static int pru_rproc_debug_read_regs(struct seq_file *s, void *data)
+{
+	struct rproc *rproc = s->private;
+	struct pru_rproc *pru = rproc->priv;
+	int i, nregs = 32;
+	int ret = 0;
+	u32 pru_sts;
+	int pru_is_running;
+
+	ret += seq_puts(s, "============== Control Registers ==============\n");
+	ret += seq_printf(s, "CTRL      := 0x%08x\n",
+			  pru_control_read_reg(pru, PRU_CTRL_CTRL));
+	pru_sts = pru_control_read_reg(pru, PRU_CTRL_STS);
+	ret += seq_printf(s, "STS (PC)  := 0x%08x (0x%08x)\n",
+			  pru_sts, pru_sts << 2);
+	ret += seq_printf(s, "WAKEUP_EN := 0x%08x\n",
+			  pru_control_read_reg(pru, PRU_CTRL_WAKEUP_EN));
+	ret += seq_printf(s, "CYCLE     := 0x%08x\n",
+			  pru_control_read_reg(pru, PRU_CTRL_CYCLE));
+	ret += seq_printf(s, "STALL     := 0x%08x\n",
+			  pru_control_read_reg(pru, PRU_CTRL_STALL));
+	ret += seq_printf(s, "CTBIR0    := 0x%08x\n",
+			  pru_control_read_reg(pru, PRU_CTRL_CTBIR0));
+	ret += seq_printf(s, "CTBIR1    := 0x%08x\n",
+			  pru_control_read_reg(pru, PRU_CTRL_CTBIR1));
+	ret += seq_printf(s, "CTPPR0    := 0x%08x\n",
+			  pru_control_read_reg(pru, PRU_CTRL_CTPPR0));
+	ret += seq_printf(s, "CTPPR1    := 0x%08x\n",
+			  pru_control_read_reg(pru, PRU_CTRL_CTPPR1));
+
+	ret += seq_puts(s, "=============== Debug Registers ===============\n");
+	pru_is_running = pru_control_read_reg(pru, PRU_CTRL_CTRL) &
+				CTRL_CTRL_RUNSTATE;
+	if (pru_is_running) {
+		ret += seq_puts(s, "PRU is executing, cannot print/access debug registers.\n");
+		return 0;
+	}
+
+	for (i = 0; i < nregs; i++) {
+		ret += seq_printf(s, "GPREG%-2d := 0x%08x\tCT_REG%-2d := 0x%08x\n",
+			    i, pru_debug_read_reg(pru, PRU_DEBUG_GPREG(i)),
+			    i, pru_debug_read_reg(pru, PRU_DEBUG_CT_REG(i)));
+	}
+
+	return 0;
+}
+
+static int pru_rproc_debug_regs_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, pru_rproc_debug_read_regs, inode->i_private);
+}
+
+static const struct file_operations pru_rproc_debug_regs_ops = {
+	.open = pru_rproc_debug_regs_open,
+	.read = seq_read,
+	.llseek	= seq_lseek,
+	.release = single_release,
+};
+
+/*
+ * Control PRU single-step mode
+ *
+ * This is a debug helper function used for controlling the single-step
+ * mode of the PRU. The PRU Debug registers are not accessible when the
+ * PRU is in RUNNING state.
+ *
+ * Writing a non-zero value sets the PRU into single-step mode irrespective
+ * of its previous state. The PRU mode is saved only on the first set into
+ * a single-step mode. Writing a non-zero value will restore the PRU into
+ * its original mode.
+ */
+static int pru_rproc_debug_ss_set(void *data, u64 val)
+{
+	struct rproc *rproc = data;
+	struct pru_rproc *pru = rproc->priv;
+	u32 reg_val;
+
+	val = val ? 1 : 0;
+	if (!val && !pru->dbg_single_step)
+		return 0;
+
+	reg_val = pru_control_read_reg(pru, PRU_CTRL_CTRL);
+
+	if (val && !pru->dbg_single_step)
+		pru->dbg_continuous = reg_val;
+
+	if (val)
+		reg_val |= CTRL_CTRL_SINGLE_STEP | CTRL_CTRL_EN;
+	else
+		reg_val = pru->dbg_continuous;
+
+	pru->dbg_single_step = val;
+	pru_control_write_reg(pru, PRU_CTRL_CTRL, reg_val);
+
+	return 0;
+}
+
+static int pru_rproc_debug_ss_get(void *data, u64 *val)
+{
+	struct rproc *rproc = data;
+	struct pru_rproc *pru = rproc->priv;
+
+	*val = pru->dbg_single_step;
+	return 0;
+}
+DEFINE_SIMPLE_ATTRIBUTE(pru_rproc_debug_ss_fops, pru_rproc_debug_ss_get,
+			pru_rproc_debug_ss_set, "%llu\n");
+
+/*
+ * Create PRU-specific debugfs entries
+ *
+ * The entries are created only if the parent remoteproc debugfs directory
+ * exists, and will be cleaned up by the remoteproc core.
+ */
+static void pru_rproc_create_debug_entries(struct rproc *rproc)
+{
+	if (!rproc->dbg_dir)
+		return;
+
+	debugfs_create_file("regs", 0400, rproc->dbg_dir,
+			    rproc, &pru_rproc_debug_regs_ops);
+	debugfs_create_file("single_step", 0600, rproc->dbg_dir,
+			    rproc, &pru_rproc_debug_ss_fops);
+}
+
 static void pruss_init_intc(struct pruss *pruss)
 {
 	int i;
@@ -699,6 +829,8 @@ static int pru_rproc_probe(struct platform_device *pdev)
 		goto put_mbox;
 	}
 
+	pru_rproc_create_debug_entries(rproc);
+
 	/*
 	 * rproc_add will boot the processor if the corresponding PRU
 	 * has a virtio device published in its resource table. If not
-- 
1.7.9.5

